src/compiler/hxb/hxbWriter.ml

```diff
@@ -0,0 +1,1465 @@
+		(* Printf.eprintf "  Write class ref %d for %s\n" i (snd c.cl_path); *)
+		chunk#write_uleb128 i
+
+	method write_enum_ref (en : tenum) =
+		let i = enums#get_or_add en.e_path en in
+		(* Printf.eprintf "  Write enum ref %d for %s\n" i (snd en.e_path); *)
+		chunk#write_uleb128 i
+
+	method write_typedef_ref (td : tdef) =
+		let i = typedefs#get_or_add td.t_path td in
+		(* Printf.eprintf "  Write typedef ref %d for %s\n" i (s_type_path td.t_path); *)
+		chunk#write_uleb128 i
+
+	method write_abstract_ref (a : tabstract) =
+		let i = abstracts#get_or_add a.a_path a in
+		(* Printf.eprintf "  Write abstract ref %d for %s\n" i (snd a.a_path); *)
+		chunk#write_uleb128 i
+
+	method write_anon_ref (an : tanon) =
+		let pfm = Option.get (anon_id#identify true (TAnon an)) in
+		let i = anons#get_or_add pfm.pfm_path (an,ttp_key) in
+		(* Printf.eprintf "  Write anon ref %d for %s\n" i (s_type_path pfm.pfm_path); *)
+		chunk#write_uleb128 i
+
+	method write_field_ref (source : field_source) (cf : tclass_field) =
+		chunk#write_string cf.cf_name
+
+	method write_enum_field_ref ef =
+		chunk#write_string ef.ef_name
+
+	(* Type instances *)
+
+	method write_type_instance t =
+		let write_function_arg (n,o,t) =
+			chunk#write_string n;
+			chunk#write_bool o;
+			self#write_type_instance t;
+		in
+		match t with
+		| TMono r ->
+			begin match r.tm_type with
+			| None ->
+				chunk#write_byte 0
+			| Some t ->
+				chunk#write_byte 1;
+				self#write_type_instance t
+			end
+		| TInst({cl_kind = KTypeParameter _} as c,[]) ->
+			begin try
+				let i = field_type_parameters#get (snd c.cl_path) in
+				chunk#write_byte 5;
+				chunk#write_uleb128 i
+			with Not_found -> try
+				let i = type_type_parameters#get (snd c.cl_path) in
+				chunk#write_byte 6;
+				chunk#write_uleb128 i
+			with Not_found ->
+				error ("Unbound type parameter " ^ (s_type_path c.cl_path))
+			end
+		| TInst(c,[]) ->
+			chunk#write_byte 10;
+			self#write_class_ref c;
+		| TEnum(en,[]) ->
+			chunk#write_byte 11;
+			self#write_enum_ref en;
+		| TType(td,[]) ->
+			chunk#write_byte 12;
+			begin match td.t_type with
+				| TAnon an when PMap.is_empty an.a_fields ->
+					chunk#write_byte 0;
+				| TAnon an ->
+					chunk#write_byte 1;
+					self#write_anon_ref an;
+				| _ ->
+					chunk#write_byte 2;
+					self#write_typedef_ref td;
+			end;
+		| TAbstract(a,[]) ->
+			chunk#write_byte 13;
+			self#write_abstract_ref a;
+		| TInst(c,tl) ->
+			chunk#write_byte 14;
+			self#write_class_ref c;
+			self#write_types tl
+		| TEnum(en,tl) ->
+			chunk#write_byte 15;
+			self#write_enum_ref en;
+			self#write_types tl
+		| TType(td,tl) ->
+			chunk#write_byte 16;
+			(* Printf.eprintf "  TType %d for %s\n" 16 (s_type_path td.t_path); *)
+			self#write_typedef_ref td;
+			self#write_types tl
+		| TAbstract(a,tl) ->
+			chunk#write_byte 17;
+			self#write_abstract_ref a;
+			self#write_types tl
+		(* TODO what to do with void special case? *)
```

Comment:
This is just a small optimization because there are many functions with Void returns. Did this cause any problems?